### ISSUE TITLE
Projection int overflow for zoom=22

### ISSUE THREAD

**Issue by lurtzn** [osmdroid/osmdroid#441]
in Projection.java line 136
`
if (absY > Math.abs(out.y - yCompare) && mMapViewHeight < 2 * mapSize) {  
               out.y -= mapSize;
          }
          if (absY > Math.abs(out.y + yCompare) || mMapViewHeight >= 2 * mapSize) { 
               out.y += mapSize;
          }`
2 * mapSize > Interger.MAX_VALUE

solution:
mMapViewHeight / 2 < mapSize 
mMapViewHeight / 2 >=  mapSize

**Comment by spyhunter99** [comment]
Zoom levels over 20 have known issues. This probably won't fix it

**Comment by lurtzn** [comment]
it works fine after you change that line

**Comment by spyhunter99** [comment]
interesting, i'll look into it. Does it "work fine' even with Markers/Icon
overlays, etc? Which part "works fine"? There's been many recent issues
about zoom > 20 but few solutions.

**Comment by spyhunter99** [comment]
can you open a pr for this (or a patch)? I'm not super clear on how to apply this

**Comment by lurtzn** [comment]
osmdroid/osmdroid-android/src/main/java/org/osmdroid/views/Projection.java
protected Point adjustForDateLine(int x, int y, Point reuse)

this function is called by public Point toPixels(final IGeoPoint in, final Point reuse)

so its basicly called by every time (afaik) when you draw any overlay

if you have a tilesize of 256 = 2^8 you have a max zoom of 23 (cause you have 31 bits = 23 + 8)

2 * mapSize wasted another bits for no reason. that is what i wanted to say.

if you divide mMapViewHeight / 2 you don't waste the bit, therefore you could zoom 1 level deper.

and after that change you can't improve the zoom anymore.

you could improve the zoom level by one if you would use unsinged int for comparing the 2 values.

but after that you would need to change to longs.

following classes would need to change then: mapview.scroller + both tileserver classes.


since you could improve the potential zoom by one level by only changing 
2 * mapSize to mMapViewHeight / 2
i wanted to tell you that. ;-D

for what i want to do i need a zoom level > 25 anyway. therefore i can't use osmdroid anyway, unless i completly change how the mapviewer.scroler works.

hope you understood what i wanted to say ;-P

**Comment by spyhunter99** [comment]
got it now. I was also looking into fixing the scroller vs canvas thing but it's more complex than it seems. Zoom greater than 20 actually has issues too.

**Comment by spyhunter99** [comment]
With and without this change, icons and polylines are still visible at zoom 21. 22 they all disappear.  

so what does this change actually fix? What's the behavioral difference?

**Comment by lurtzn** [comment]
they shouldn't disappear.

**Comment by spyhunter99** [comment]
haha yes i know they shouldn't. Do you have different behavior?

**Comment by lurtzn** [comment]
yes. they don't disappear on my app after i changed the 2 lines.

**Comment by spyhunter99** [comment]
What geographic area of the world are you testing with?

**Comment by lurtzn** [comment]
51.7875, 6.135278 = kleve 
49.783333, 9.933333 = wuerzburg

**Comment by spyhunter99** [comment]
Update, i reran the tests at your coordinates
Marker class works ok up to and including zoom 22.
Polyline works up to and including 21, but disappears at 22
ItemizedIconOverlay works ok up to and including zoom 22.

I then reverted the change and had the same result.

i also tested with NYC sample with and without the change. no delta. I just pushed the change and the sample update. Feel free to test and replicate on your end
